flink-kubernetes/src/test/java/org/apache/flink/kubernetes/highavailability/KubernetesLeaderElectionServiceTest.java

```diff
@@ -0,0 +1,146 @@
+					assertThat(getLeaderConfigMap().getData().size(), is(0));
+				});
+		}};
+	}
+
+	@Test
+	public void testLeaderConfigMapDeletedExternally() throws Exception {
+		new Context() {{
+			runTestAndGrantLeadershipToContender(
+				() -> {
+					final FlinkKubeClient.WatchCallbackHandler<KubernetesConfigMap> callbackHandler =
+						configMapsAndDoCallbackFuture.get(TIMEOUT, TimeUnit.MILLISECONDS);
+					assertThat(callbackHandler, is(notNullValue()));
+
+					callbackHandler.onDeleted(Collections.singletonList(configMapStore.remove(LEADER_CONFIGMAP_NAME)));
+					// The ConfigMap should be created again.
+					assertThat(getLeaderConfigMap().getData().get(Constants.LEADER_ADDRESS_KEY), is(LEADER_URL));
+				});
+		}};
+	}
+
+	@Test
+	public void testLeaderConfigMapUpdatedExternally() throws Exception {
+		new Context() {{
+			runTestAndGrantLeadershipToContender(
+				() -> {
+					final FlinkKubeClient.WatchCallbackHandler<KubernetesConfigMap> callbackHandler =
+						configMapsAndDoCallbackFuture.get(TIMEOUT, TimeUnit.MILLISECONDS);
+					assertThat(callbackHandler, is(notNullValue()));
+					// Update ConfigMap with wrong data
+					getLeaderConfigMap().getData().put(Constants.LEADER_ADDRESS_KEY, "wrong data");
+					callbackHandler.onModified(Collections.singletonList(configMapStore.get(LEADER_CONFIGMAP_NAME)));
+					// The ConfigMap should be corrected
+					assertThat(getLeaderConfigMap().getData().get(Constants.LEADER_ADDRESS_KEY), is(LEADER_URL));
+				});
+		}};
+	}
+
+	/**
+	 * Start multiple leaders, if current leader dies, a new one could take over and update the leader
+	 * information successfully.
+	 */
+	@Test
+	public void testMultipleLeaders() throws Exception {
+		final int leaderNum = 3;
+		final AtomicBoolean[] leaderController = new AtomicBoolean[leaderNum];
+		final TestingContender[] contenders = new TestingContender[leaderNum];
+		final KubernetesLeaderElectionService[] leaderElectionServices = new KubernetesLeaderElectionService[leaderNum];
+		for (int i = 0; i < leaderNum; i++) {
+			leaderController[i] = new AtomicBoolean(false);
+			leaderElectionServices[i] = new Context().createLeaderElectionService(leaderController[i]);
+			contenders[i] = new TestingContender(getLeaderUrl(i), leaderElectionServices[i]);
+			leaderElectionServices[i].start(contenders[i]);
+		}
+		leaderController[0].set(true);
+		contenders[0].waitForLeader(TIMEOUT);
+		assertThat(
+			configMapStore.get(LEADER_CONFIGMAP_NAME).getData().get(Constants.LEADER_ADDRESS_KEY),
+			is(getLeaderUrl(0)));
+		// Leader 0 died
+		leaderController[0].set(false);
+		contenders[0].waitForRevokeLeader(TIMEOUT);
+		// Leader 2 try to acquire
+		leaderController[2].set(true);
+		contenders[2].waitForLeader(TIMEOUT);
+		assertThat(
+			configMapStore.get(LEADER_CONFIGMAP_NAME).getData().get(Constants.LEADER_ADDRESS_KEY),
+			is(getLeaderUrl(2)));
+		for (int i = 0; i < leaderNum; i++) {
+			leaderElectionServices[i].stop();
+		}
+	}
```

Comment:
Shouldn't we also have a test were we have multiple leaders concurrently trying to obtain the leadership using the proper `KubernetesLeaderElector`?

Reply:
It is hard to use the real `KubernetesLeaderElector` in the UT. That's why I add a `MockKubernetesLeaderElector` in the `TestingFlinkKubeClient` to valid the contract tests for leader election/retrieval.

Reply:
I have added a ITCase for the concurrent leader election.